### PR TITLE
fix(app):  Introduced flag to specify leader election behavious

### PR DESCRIPTION
**Pull request by CharlesQQ**
resolve: https://github.com/volcano-sh/volcano/issues/2798

### CODE REVIEW

**Comment by lowang-bh** [comment]
I think we should do some checks between those three parameters

**Comment by CharlesQQ** [comment]
cc @wangyang0616

**Comment by wangyang0616** [comment]
I agree with @lowang-bh is suggestion that the input parameters need to be verified.

**Comment by CharlesQQ** [comment]
> 
@lowang-bh @wangyang0616 
there have the checks for three parameter in leaderelection package, so no need to check again.
image

**Comment by CharlesQQ** [comment]
cc @wangyang0616

File: cmd/controller-manager/app/options/options.go [hunk 1]
```diff
@@ -83,6 +93,19 @@ func (s *ServerOption) AddFlags(fs *pflag.FlagSet) {
 	fs.BoolVar(&s.EnableLeaderElection, "leader-elect", true, "Start a leader election client and gain leadership before "+
 		"executing the main loop. Enable this when running replicated vc-controller-manager for high availability; it is enabled by default")
 	fs.StringVar(&s.LockObjectNamespace, "lock-object-namespace", defaultLockObjectNamespace, "Define the namespace of the lock object; it is volcano-system by default")
+	fs.DurationVar(&s.LeaseDuration, "leader-elect-lease-duration", defaultElectionLeaseDuration, ""+
```
Comment (wangyang0616) [inline cmd/controller-manager/app/options/options.go:96]: I'm thinking here mainly about the string here, whether to comply with the 120-character requirement, or to support legibility, is it easier to read if the two lines of text description information are closed?

Comment (CharlesQQ) [reply cmd/controller-manager/app/options/options.go:96]: > I'm thinking here mainly about the string here, whether to comply with the 120-character requirement, or to support legibility, is it easier to read if the two lines of text description information are closed?

It's same as kubernetes comments, sames that comply with the 120-character requirement
https://github.com/kubernetes/kubernetes/blob/ea30d100f6b13a47b4153362698b89256077776c/staging/src/k8s.io/component-base/config/options/leaderelectionconfig.go#L30C1-L42

File: cmd/controller-manager/app/options/options.go [hunk 2]
```diff
@@ -47,6 +54,9 @@ type ServerOption struct {
 	CaCertData           []byte
 	EnableLeaderElection bool
```
Comment (Monokaix) [inline cmd/controller-manager/app/options/options.go:55]: Leader election related parameters can be placed in an independent flagset, please refer to kubernets to make it more graceful.

Comment (CharlesQQ) [reply cmd/controller-manager/app/options/options.go:55]: @Monokaix updated!  pls  check again.

File: cmd/scheduler/app/options/options.go [hunk 3]
```diff
@@ -195,10 +192,10 @@ func (s *ServerOption) ParseCAFiles(decryptFunc DecryptFunc) error {
 	return nil
 }
 
-// Default new and registry a default one
```
Comment (Monokaix) [inline cmd/scheduler/app/options/options.go:198]: Why is this commented?

File: pkg/util/util.go [hunk 4]
```diff
@@ -32,3 +47,31 @@ func GenerateSchedulerName(schedulerNames []string) string {
 
 	return defaultSchedulerName
 }
+
+// BindLeaderElectionFlags binds the LeaderElectionConfiguration struct fields to a flagset
+func BindLeaderElectionFlags(l *config.LeaderElectionConfiguration, fs *pflag.FlagSet) {
```
Comment (Monokaix) [inline pkg/util/util.go:52]: How about import `k8s.io/component-base/config/options/leaderelectionconfig.go` directly?

Comment (CharlesQQ) [reply pkg/util/util.go:52]: > How about import k8s.io/component-base/config/options/leaderelectionconfig.go directly?
en.., I want set default value in function `BindLeaderElectionFlags`; if not, must add a function to set LeaderElectionConfiguration  default values.

File: cmd/controller-manager/app/options/options.go [hunk 5]
```diff
@@ -38,16 +38,16 @@ const (
...
 // ServerOption is the main context object for the controllers.
 type ServerOption struct {
-	KubeClientOptions    kube.ClientOptions
-	CertFile             string
-	KeyFile              string
-	CaCertFile           string
-	CertData             []byte
-	KeyData              []byte
-	CaCertData           []byte
-	EnableLeaderElection bool
```
Comment (lowang-bh) [inline cmd/controller-manager/app/options/options.go:48]: How about keeping the old one for better compatibility?

File: cmd/scheduler/app/options/options.go (hunk 3 above)
Comment (CharlesQQ) [reply cmd/scheduler/app/options/options.go:198]: > Why is this commented?
useless code, I will delete that.

File: cmd/scheduler/app/options/options.go [hunk 6]
```diff
@@ -195,10 +192,10 @@ func (s *ServerOption) ParseCAFiles(decryptFunc DecryptFunc) error {
...
-func Default() *ServerOption {
-	s := NewServerOption()
-	s.AddFlags(pflag.CommandLine)
-	s.RegisterOptions()
-	return s
-}
+//// Default new and registry a default one
+//func Default() *ServerOption {
```
Comment (lowang-bh) [inline cmd/scheduler/app/options/options.go:196]: Please keep it. It is used in other test packages.

File: pkg/util/util.go (hunk 4 above)
Comment (Monokaix) [reply pkg/util/util.go:52]: But actually it's a copy from k8s: )

File: cmd/controller-manager/app/options/options.go [hunk 7]
```diff
@@ -18,9 +18,9 @@ package options
 
 import (
 	"fmt"
-	"os"
-
 	"github.com/spf13/pflag"
+	"k8s.io/component-base/config"
+	"os"
```
Comment (Monokaix) [inline cmd/controller-manager/app/options/options.go:23]: Plz sort imports.

File: cmd/controller-manager/app/options/options.go (hunk 5 above)
Comment (Monokaix) [reply cmd/controller-manager/app/options/options.go:48]: `leader-elect` flag is kept when using BindLeaderElectionFlags.

Comment (CharlesQQ) [reply cmd/controller-manager/app/options/options.go:48]: > How about keeping the old one for better compatibility?
Use leaderElection flagSet instead, it's a leaderElectionOption field.

File: cmd/scheduler/app/options/options.go (hunk 6 above)
Comment (CharlesQQ) [reply cmd/scheduler/app/options/options.go]: > Please keep it. It is used in other test packages.
all right!

File: cmd/controller-manager/app/options/options.go (hunk 7 above)
Comment (CharlesQQ) [reply cmd/controller-manager/app/options/options.go]: > Plz sort imports.
done!

File: cmd/controller-manager/app/options/options.go (hunk 5 above)
Comment (Monokaix) [reply cmd/controller-manager/app/options/options.go:48]: But `lock-object-namespace` flag seems has been removed.

Comment (CharlesQQ) [reply cmd/controller-manager/app/options/options.go:48]: > But lock-object-namespace flag seems has been removed.

use the `leader-elect-resource-namespace`  flag instead

Comment (Monokaix) [reply cmd/controller-manager/app/options/options.go:48]: So plz also update params in `helm` : )

File: pkg/util/util.go [hunk 8]
```diff
@@ -1,7 +1,21 @@
 package util
 
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/tools/leaderelection/resourcelock"
+	"k8s.io/component-base/config"
+	"time"
```
Comment (Monokaix) [inline pkg/util/util.go:7]: sort imports.

File: cmd/scheduler/app/options/options.go [hunk 9]
```diff
@@ -147,7 +144,7 @@ func (s *ServerOption) AddFlags(fs *pflag.FlagSet) {
 
 // CheckOptionOrDie check lock-object-namespace when LeaderElection is enabled.
 func (s *ServerOption) CheckOptionOrDie() error {
-	if s.EnableLeaderElection && s.LockObjectNamespace == "" {
+	if s.LeaderElection.LeaderElect && s.LeaderElection.ResourceLock == "" {
 		return fmt.Errorf("lock-object-namespace must not be nil when LeaderElection is enabled")
```
Comment (Monokaix) [inline cmd/scheduler/app/options/options.go:148]: Change the description too.

File: cmd/controller-manager/app/options/options.go (hunk 5 above)
Comment (Monokaix) [reply cmd/controller-manager/app/options/options.go:48]: I think we should better keep the old flag and mark it deprecated first.

File: cmd/controller-manager/app/options/options.go [hunk 10]
```diff
@@ -102,7 +100,7 @@ func (s *ServerOption) AddFlags(fs *pflag.FlagSet) {
 
 // CheckOptionOrDie checks the LockObjectNamespace.
 func (s *ServerOption) CheckOptionOrDie() error {
-	if s.EnableLeaderElection && s.LockObjectNamespace == "" {
+	if s.LeaderElection.LeaderElect && s.LeaderElection.ResourceLock == "" {
```
Comment (Monokaix) [inline cmd/controller-manager/app/options/options.go:103]: Here we can call `ValidateLeaderElectionConfiguration` to check all params.

File: cmd/controller-manager/app/options/options.go (hunk 5 above)
Comment (CharlesQQ) [reply cmd/controller-manager/app/options/options.go:48]: > So plz also update params in helm  I think we should better keep the old flag and mark it deprecated first.

sure!

File: cmd/controller-manager/app/options/options.go (hunk 10 above)
Comment (CharlesQQ) [reply cmd/controller-manager/app/options/options.go]: All that is updated!

File: cmd/controller-manager/main.go [hunk 11]
```diff
@@ -46,8 +46,12 @@ func main() {
 	runtime.GOMAXPROCS(runtime.NumCPU())
 	klog.InitFlags(nil)
 
+	fs := pflag.CommandLine
 	s := options.NewServerOption()
-	s.AddFlags(pflag.CommandLine)
+
+	s.AddFlags(fs)
+	commonutil.LeaderElectionDefault(&s.LeaderElection)
+	componentbaseoptions.BindLeaderElectionFlags(&s.LeaderElection, fs)
```
Comment (Monokaix) [inline cmd/controller-manager/main.go:55]: Just mark LockObjectNamespace deprecated is not enough, we should set `ResourceNamespace` value to LockObjectNamespace  when it's not empty: )

Comment (CharlesQQ) [reply cmd/controller-manager/main.go:58]: > Just mark LockObjectNamespace deprecated is not enough, we should set ResourceNamespace value to LockObjectNamespace when it's not empty: )

All right！

File: cmd/controller-manager/app/options/options.go (hunk 5 above)
Comment (lowang-bh) [reply cmd/controller-manager/app/options/options.go:48]: > I think we should better keep the old flag and mark it deprecated first.

I also means keeping the old one and marking it is deprecated in recent release.

File: cmd/controller-manager/app/options/options.go [hunk 12]
```diff
@@ -84,9 +89,7 @@ func (s *ServerOption) AddFlags(fs *pflag.FlagSet) {
 		"File containing the default x509 Certificate for HTTPS. (CA cert, if any, concatenated "+
 		"after server cert).")
 	fs.StringVar(&s.KeyFile, "tls-private-key-file", s.KeyFile, "File containing the default x509 private key matching --tls-cert-file.")
-	fs.BoolVar(&s.EnableLeaderElection, "leader-elect", true, "Start a leader election client and gain leadership before "+
```
Comment (lowang-bh) [inline cmd/controller-manager/app/options/options.go:87]: why this is removed?

File: cmd/scheduler/app/options/options.go [hunk 13]
```diff
@@ -107,16 +112,13 @@ func (s *ServerOption) AddFlags(fs *pflag.FlagSet) {
 		"File containing the default x509 Certificate for HTTPS. (CA cert, if any, concatenated "+
 		"after server cert).")
 	fs.StringVar(&s.KeyFile, "tls-private-key-file", s.KeyFile, "File containing the default x509 private key matching --tls-cert-file.")
+	fs.StringVar(&s.LockObjectNamespace, "lock-object-namespace", defaultLockObjectNamespace, "Define the namespace of the lock object; it is volcano-system by default, will deprecated, please use --leader-elect-resource-namespace instead.")
 	// volcano scheduler will ignore pods with scheduler names other than specified with the option
 	fs.StringArrayVar(&s.SchedulerNames, "scheduler-name", []string{defaultSchedulerName}, "vc-scheduler will handle pods whose .spec.SchedulerName is same as scheduler-name")
 	fs.StringVar(&s.SchedulerConf, "scheduler-conf", "", "The absolute path of scheduler configuration file")
 	fs.DurationVar(&s.SchedulePeriod, "schedule-period", defaultSchedulerPeriod, "The period between each scheduling cycle")
 	fs.StringVar(&s.DefaultQueue, "default-queue", defaultQueue, "The default queue name of the job")
-	fs.BoolVar(&s.EnableLeaderElection, "leader-elect", true,
```
Comment (lowang-bh) [inline cmd/scheduler/app/options/options.go:115]: why this is removed, how about deprecated？

Comment (Monokaix) [reply cmd/scheduler/app/options/options.go:115]: New flagset contains this flag, so it's not removed autually.

**Review by william-wang — APPROVED** [review]
/approve

**Pull request merged**